File: solutions/haskell/sum-of-multiples/1/src/SumOfMultiples.hs
module SumOfMultiples
    (
      sumOfMultiples
    , sumOfMultiplesDefault
    ) where

import Data.List (nub)

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples ns upTo = sum $ nub $ concatMap multiples ns
    where multiples n = [(n*x) | x <- [1..upTo], (n*x) < upTo]

sumOfMultiplesDefault :: Integer -> Integer
sumOfMultiplesDefault = sumOfMultiples [3, 5]

File: solutions/ocaml/sum-of-multiples/1/sum_of_multiples.ml
let multiples limit = function
  | 0 -> []
  | base -> List.init (pred limit / base) (fun i -> succ i * base)
let sum bases limit =
  List.concat_map (multiples limit) bases
  |> List.sort_uniq Int.compare 
  |> List.fold_left ( + ) 0